lib/RecipeExecutor.jsm

```diff
@@ -0,0 +1,1075 @@
+ *
+ * Recipes are sandboxed by the fact that the step functions must be explicitly
+ * whitelisted. Functions whitelisted for builder recipes are specifed in the
+ * RecipeExecutor.ITEM_BUILDER_REGISTRY, while combiner functions are whitelisted
+ * in RecipeExecutor.ITEM_COMBINER_REGISTRY .
+ */
+this.RecipeExecutor = class RecipeExecutor {
+  constructor(nbTaggers, nmfTaggers) {
+    this.ITEM_BUILDER_REGISTRY = {
+      nb_tag: this.naiveBayesTag,
+      conditionally_nmf_tag: this.conditionallyNmfTag,
+      accept_item_by_field_value: this.acceptItemByFieldValue,
+      tokenize_url: this.tokenizeUrl,
+      get_url_domain: this.getUrlDomain,
+      tokenize_field: this.tokenizeField,
+      copy_value: this.copyValue,
+      keep_top_k: this.keepTopK,
+      scalar_multiply: this.scalarMultiply,
+      elementwise_multiply: this.elementwiseMultiply,
+      vector_multiply: this.vectorMultiply,
+      scalar_add: this.scalarAdd,
+      vector_add: this.vectorAdd,
+      make_boolean: this.makeBoolean,
+      whitelist_fields: this.whitelistFields,
+      filter_by_value: this.filterByValue,
+      l2_normalize: this.l2Normalize,
+      prob_normalize: this.probNormalize,
+      set_default: this.setDefault,
+      lookupValue: this.lookupValue,
+      copy_to_map: this.copyToMap,
+      scalar_multiply_tag: this.scalarMultiplyTag,
+      apply_softmax_tags: this.applySoftmaxTags
+    };
+    this.ITEM_COMBINER_REGISTRY = {
+      combiner_add: this.combinerAdd,
+      combiner_max: this.combinerMax,
+      combiner_collect_values: this.combinerCollectValues
+    };
+    this.nbTaggers = nbTaggers;
+    this.nmfTaggers = nmfTaggers;
+  }
+
+  /**
+   * Determines the type of a field. Valid types are:
+   *   string
+   *   number
+   *   array
+   *   map (strings to anything)
+   */
+  _typeOf(data) {
+    let t = typeof(data);
+    if (t === "object") {
+      if (Array.isArray(data)) {
+        return "array";
+      }
+      return "map";
+    }
+    return t;
+  }
+
+  /**
+   * Returns returns a scalar, either by because it was a constant, or by
+   * looking it up from the item. Allows for a default value if the lookup
+   * fails.
+   */
+  _lookupScalar(item, k, dfault) {
+    if (this._typeOf(k) === "number") {
+      return k;
+    } else if ((this._typeOf(k) === "string") && (k in item) && (this._typeOf(item[k]) === "number")) {
+      return item[k];
+    }
+    return dfault;
+  }
+
+  /**
+   * Simply appends all the strings from a set fields together. If the field
+   * is a list, then the cells of the list are append.
+   */
+  _assembleText(item, fields) {
+    let textArr = [];
+    for (let field of fields) {
+      if (field in item) {
+        let type = this._typeOf(item[field]);
+        if (type === "string") {
+          textArr.push(item[field]);
+        } else if (type === "array") {
+          for (let ele of item[field]) {
+            textArr.push(ele);
```

Comment:
Do you want to check the type of `ele`, only add it if it's a string?

Reply:
we'll just convert it to a string.